### ISSUE TITLE
Rtags with builds inside a Docker

### ISSUE THREAD

**Issue by nickjmeyer** [Andersbakken/rtags#1085]
### Please mark appropriate
 - [x] rtags (rdm/rc/rp)
 - Emacs Lisp
   - [ ] ac-rtags.el
   - [ ] company-rtags.el
   - [ ] helm-rtags.el
   - [ ] flycheck-rtags.el
   - [ ] ivy-rtags.el
   - [ ] rtags.el

### Problem description
One of the projects I work on builds everything in a docker container.  Thus all the compile commands have paths relative to the container and not the actual disk.  Is there a way to use rtags in this scenario that doesn't require me running emacs and rtags inside the docker?

**Comment by pedro-esteves-pinto** [comment]
Interestingly enough, I do exactly that, i.e. run rtags and emacs inside the container and this is pretty seamless. If that doesn't work for you, could you perhaps mount your HOME directory inside the container?

**Comment by nickjmeyer** [comment]
Mounting the home directory would make it much more convenient.  That way I don't need two emacs instances.

Though, I'm wondering, would it be possible to have rtags run inside a docker?  Customize the emacs bindings so any rtags call actually runs rtags in a docker.  That way I could have it run inside my docker build image.  This would allow you to run emacs as normal.

Do you think that's possible?  I'm going to try to dig in and see if I can get something working.  Ideally, you'd be able to indicate the "wrapper command" in a `.dir-locals.el` file.

**Comment by pedro-esteves-pinto** [comment]
Yes, its definitely possible to run rtags inside docker - but in that case I think you would start docker as an interactive long-living session and just run rdm inside. I run emacs in there as well but you wouldn't have to do that. Here's my docker launch command:

``#!/bin/bash
nohup docker run \
      --rm \
      --net host \
      -h $(hostname)_docker \
      -v /etc/passwd:/etc/passwd \
      --security-opt seccomp=unconfined \
      -v /var/log/syslog:/var/log/syslog \
      -v  /etc/group:/etc/group \
      -v /tmp/.X11-unix:/tmp/.X11-unix \
      -u $UID \
      -v /home:/home \
      -e DISPLAY=$DISPLAY \
      -t my_container \
      sh -c 'rdm --daemon -L ~/temp/rtags_log & emacs' &

**Comment by nickjmeyer** [comment]
I would have to also run `rc` inside the docker right?

**Comment by nickjmeyer** [comment]
Thanks for passing along the docker command.  Much appreciated.

If I only run rtags in the docker, I'll probably have to expose the port right?  I don't see that in your docker command, but that makes sense because you have emacs inside the docker as well.

I'm guessing I'll also have to run `rc` in a docker too right?

I'll give this a shot some time this week.

**Comment by pedro-esteves-pinto** [comment]
If the file system is shared then your paths will all align and I don't think you would absolutely need to run anithing inside docker. However if you want to make sure the clang version is the same inside and outside the container and that all your system libraries are the same then you could probably just run rdm inside and rc outside, exposing the rdm port in the docker command line as you mention.

**Comment by nicber** [comment]
I have the same problem at work.
I solved it with a script that copies the compile_commands.json file from the docker filesystem, patches the paths in it and then calls rc.
I ran `clang++4.0 -v -E -dM a_test_file.cpp` inside the docker container, then I pass the system include paths to rdm alongside the `--no-libclang-includes` flag. I need to mount the docker container filesystem in a directory using sshfs and then change each system include by prefixing it with the directory where I mounted the container's filesystem.

It is quite a lot of work, but I can run rdm and emacs on the host machine.

**Comment by mcraveiro** [comment]
Hi guys,

joining the discussion a bit late, but got exactly the same problem as described by @nickjmeyer. I quite like the idea of mounting home in the docker container, but I'd like to keep rtags also within the container so that everyone has exactly the same setup. This means we do not have to install any dev tools at all in the host.

As far as I can see from the discussion above, the only snag is just rc. Presumably one would have to copy rc outside of the container and expose the port right? I will give this a go and report back.

**Comment by nicber** [comment]
rdm creates the socket file inside the home dir, so if you mount the home dir directly in the container it should be in the correct location for rc to find it.

**Comment by samwise** [comment]
Hi Mario, 

I have gone to a few iterations of this type of setup. If you do not mind running emacs in the container that is the simplest, most deterministic way to go about it, you just run rc, rtags, etc all from the same environment and everything is ready to go with as soon as you enter the container. If you want to run emacs outside, then I think you have to run rc  outside as well, or at least with out of the box rtags I wasn't able to persuade emacs to use the rc inside the container. As you mention you also have to trigger the build inside the container from emacs, I use something like this:

cid=`docker ps | grep <my_docker_image>| awk '{print $1}'`
docker exec $cid /bin/bash -c "cd  <your_build_dir> && make"
 
ssh and tramp probably work as well.

**Comment by mcraveiro** [comment]
Thanks @nicber, this is all sounding quite promising. And now I found docker-tramp [1] too, so no need for SSH methinks

[1] https://github.com/emacs-pe/docker-tramp.el

**Comment by nmeyer-argo** [comment]
Well I finally got around to trying this out.  I'm running `rdm` and `rc` inside the docker I use to build my code.  When I initialize the database, everything runs fine.  However, if I restart rdm I see a bunch of errors that look like `FileSystemWatcher::watch() watch failed for <file_name>`.

Anyone have an idea why I might see that error?

**Comment by nickjmeyer** [comment]
Well I finally got around to trying this out. I'm running rdm and rc inside the docker I use to build my code. When I initialize the database, everything runs fine. However, if I restart rdm I see a bunch of errors that look like FileSystemWatcher::watch() watch failed for <file_name>.

Anyone have an idea why I might see that error?

**Comment by nickjmeyer** [comment]
I posted another issue #1143 for this error.
